mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,1129 @@
+            if should_not_overwrite {
+                return Err(StorageError::RingBuffer(RingBufferError::Full));
+            }
+        }
+
+        let should_flush = self.should_flush();
+
+        save_block_header(
+            &mut self.mmap,
+            &block_header,
+            start,
+            self.max_file_size,
+            should_flush,
+        )?;
+
+        let mut end = start + block_size;
+        start = end % self.max_file_size;
+
+        save_data(
+            &mut self.mmap,
+            &data,
+            start,
+            self.max_file_size,
+            should_flush,
+        )?;
+
+        end = start + data_size;
+        self.pointers.write = end % self.max_file_size;
+
+        // should only happen if we wrap around
+        if self.pointers.write == self.pointers.read_begin {
+            self.can_read_from_wrap_around = true;
+        }
+
+        self.wake_up_task();
+
+        self.order += 1;
+
+        self.flush_state_update(should_flush, 1, total_size, timer.elapsed());
+
+        Ok(Key { offset: key as u64 })
+    }
+
+    fn batch(&mut self, count: usize) -> StorageResult<VecDeque<(Key, Publication)>> {
+        let write_index = self.pointers.write;
+        let read_index = self.pointers.read_begin;
+        let block_size = *SERIALIZED_BLOCK_SIZE;
+
+        // If read would go into where writes are happening then we don't have data to read.
+        if self.pointers.read_end == write_index && !self.can_read_from_wrap_around {
+            return Ok(VecDeque::new());
+        }
+
+        // If we are reading then we can reset wrap around flag
+        self.can_read_from_wrap_around = false;
+
+        let mut start = read_index;
+        let mut vdata = VecDeque::new();
+        for _ in 0..count {
+            let end = start + block_size;
+            let bytes = &self.mmap[start..end];
+            // this is unused memory
+            if bytes.iter().map(|&x| x as usize).sum::<usize>() == 0 {
+                break;
+            }
+
+            let block = binary_deserialize::<BlockHeaderWithHash>(bytes)?;
+            // this means we read bytes that don't make a block, this is
+            // a really bad state to be in as somehow the pointers don't
+            // match to where data really is at.
+            if block.inner().hint() != BLOCK_HINT {
+                return Err(RingBufferError::Validate(BlockError::Hint).into());
+            }
+
+            let inner_block = block.inner();
+            let data_size = inner_block.data_size();
+            let index = inner_block.write_index();
+            start += block_size;
+            let end = start + data_size;
+            let data = load_data(&self.mmap, start, data_size, self.max_file_size)?;
+            start = end % self.max_file_size;
+            self.pointers.read_end = start;
+
+            validate(&block, &binary_serialize(&data)?)?;
```

Comment:
Maybe we can avoid unnecessary serialization here. But this is something we can look into a bit later.